### PR TITLE
fix article removal from disk cache

### PR DESCRIPTION
**Pull request by joewalsh**

### CODE REVIEW

File: Wikipedia/Code/WMFDatabaseHouseKeeper.swift [hunk 1]
```diff
@@ -109,23 +111,31 @@ import Foundation
         */
         
         let articlesToDeleteFetchRequest = WMFArticle.fetchRequest()
-        var articlesToDeletePredicate = NSPredicate(format: "viewedDate == NULL && savedDate == NULL && placesSortOrder == NULL && isExcludedFromFeed == %@", NSNumber(booleanLiteral: false))
+        var articlesToDeletePredicate = NSPredicate(format: "viewedDate == NULL && savedDate == NULL && placesSortOrder == 0 && isExcludedFromFeed == FALSE")
```
Comment (amrox) [inline Wikipedia/Code/WMFDatabaseHouseKeeper.swift]: You're sure that `FALSE` works? I did it the only way due to the way it was in ObjC and googling around, but that's obviously much simpler if it has the same effect

File: Wikipedia/Code/WMFDatabaseHouseKeeper.swift [hunk 2]
```diff
@@ -2,9 +2,11 @@ import Foundation
 
 @objc class WMFDatabaseHouseKeeper : NSObject {
     
-    @objc func performHouseKeepingOnManagedObjectContext(_ moc: NSManagedObjectContext)  throws {
+    // Returns deleted URLs
+    @objc func performHouseKeepingOnManagedObjectContext(_ moc: NSManagedObjectContext) throws -> [URL] {
```
Comment (amrox) [inline Wikipedia/Code/WMFDatabaseHouseKeeper.swift]: Just a note for future - if this object grows any further I'd like to refactor it to use a command pattern (or something).

File: Wikipedia/Code/WMFDatabaseHouseKeeper.swift (hunk 1 above)
Comment (joewalsh) [inline Wikipedia/Code/WMFDatabaseHouseKeeper.swift]: Yeah, I wasn't sure either - I tested it when I made the change, but will re-test to verify. I also read up on here: https://developer.apple.com/library/content/documentation/Cocoa/Conceptual/Predicates/Articles/pSyntax.html

**Comment by joewalsh** [comment]
Adding WIP pending addition of migration that removes articles that should have been deleted previously.

File: Wikipedia/Code/MWKDataStore.m [hunk 3]
```diff
@@ -988,7 +987,59 @@ - (NSArray *)imageInfoForArticleWithURL:(NSURL *)url {
...
+    NSError *fetchError = nil;
+    NSArray *arrayOfAllValidArticles = [self.viewContext executeFetchRequest:allValidArticleKeysFetchRequest error:&fetchError];
+
+    if (fetchError) {
+        failure(fetchError);
+        return;
+    }
+
+    if (arrayOfAllValidArticles.count == 0) {
+        [[NSFileManager defaultManager] removeItemAtPath:[self pathForSites] error:nil];
```
Comment (amrox) [inline Wikipedia/Code/MWKDataStore.m]: Is this just an optimization, or does removing the whole sites dir have some added effect?

Comment (joewalsh) [inline Wikipedia/Code/MWKDataStore.m]: Just an optimization, rather than adding them all to the to-be-deleted list

Comment (joewalsh) [inline Wikipedia/Code/MWKDataStore.m]: Might be good to move this onto the background queue though in case it hangs long enough to cause an `8badf00d`

**Comment by joewalsh** [comment]
test this please

**Pull request merged**